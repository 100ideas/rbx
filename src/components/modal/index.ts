import "./modal.sass";

export { default } from "./Modal";
