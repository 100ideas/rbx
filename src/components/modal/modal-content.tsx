import classNames from "classnames";
import React from "react";

import { forwardRefAs, HelpersProps, transformHelpers } from "../../base";

export type ModalContentModifierProps = Partial<{ className: string }>;

export type ModalContentProps = HelpersProps & ModalContentModifierProps;

export const ModalContent = forwardRefAs<ModalContentProps, "div">(
  (props, ref) => {
    const { as, ...rest } = transformHelpers(props);
    rest.className = classNames("modal-content", rest.className);
    return React.createElement(as!, { ref, ...rest });
  },
  { as: "div" },
);
