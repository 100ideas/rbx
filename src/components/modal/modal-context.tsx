import React from "react";

import { noop } from "../../utils";

export interface ModalContextValue {
  close: () => void;
  closeOnBlur: boolean;
  closeOnEsc: boolean;
}

export const initialValue: ModalContextValue = {
  close: noop,
  closeOnBlur: false,
  closeOnEsc: true,
};

export const ModalContext = React.createContext<ModalContextValue>(
  initialValue,
);
