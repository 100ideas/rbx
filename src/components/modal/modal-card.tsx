import { cx } from "emotion";
import React from "react";

import { Element } from "@/components/element";
import { ModifierProps } from "@/modifiers";
import { ModalCardBody } from "./modal-card-body";
import { ModalCardFoot } from "./modal-card-foot";
import { ModalCardHead } from "./modal-card-head";
import { ModalCardTitle } from "./modal-card-title";

export type ModalCardModifierProps = Partial<{
  children: React.ReactNode;
  className: string;
  onClose: () => void;
  style: React.CSSProperties;
}>;

export type ModalCardProps = ModifierProps & ModalCardModifierProps;

export const ModalCard = Object.assign(
  React.forwardRef<HTMLDivElement, ModalCardProps>(
    ({ className, onClose, children, ...props }, ref) => (
      <Element {...props} ref={ref} className={cx("modal-card", className)}>
        {children}
      </Element>
    ),
  ),
  {
    Body: ModalCardBody,
    Foot: ModalCardFoot,
    Head: ModalCardHead,
    Title: ModalCardTitle,
  },
);
ModalCard.defaultProps = Object.assign(
  {
    children: null,
  },
  ModalCard.defaultProps,
);
