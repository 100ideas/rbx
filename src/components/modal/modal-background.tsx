import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic } from "../../base";
import { HelpersProps } from "../../base/helpers";
import { ModalContext, ModalContextValue } from "./modal-context";

export type ModalBackgroundModifierProps = Partial<{
  onClick: React.MouseEventHandler;
}>;

export type ModalBackgroundProps = HelpersProps & ModalBackgroundModifierProps;

const onClickHandler = (
  onClick: ModalBackgroundProps["onClick"],
  ctx: ModalContextValue,
) => (event: React.MouseEvent) => {
  if (onClick !== undefined) {
    onClick(event);
  }
  if (ctx.closeOnBlur) {
    ctx.close();
  }
};

export const ModalBackground = forwardRefAs<ModalBackgroundProps>(
  ({ className, onClick, ...rest }, ref) => (
    <ModalContext.Consumer>
      {ctx => {
        const htmlProps = {
          onClick: onClickHandler(onClick, ctx),
          role: "presentation",
        };

        return (
          <Generic
            className={classNames("modal-background", className)}
            ref={ref}
            {...htmlProps}
            {...rest}
          />
        );
      }}
    </ModalContext.Consumer>
  ),
  { as: "div" },
);

ModalBackground.displayName = "Modal.Background";
ModalBackground.propTypes = {
  onClick: PropTypes.func,
};
