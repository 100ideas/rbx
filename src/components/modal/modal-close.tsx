import classNames from "classnames";
import React from "react";

import { forwardRefAs, HelpersProps, transformHelpers } from "../../base";
import { ModalContext } from "./modal-context";

export type ModalCloseModifierProps = Partial<{
  className: string;
  onClick: React.MouseEventHandler<any>;
}>;

export type ModalCloseProps = HelpersProps & ModalCloseModifierProps;

export const ModalClose = forwardRefAs<ModalCloseProps, "button">(
  (props, ref) => {
    const { as, onClick, ...rest } = transformHelpers(props);
    rest.className = classNames("modal-close", "is-large", rest.className);
    return (
      <ModalContext.Consumer>
        {({ onClose }) =>
          React.createElement(as!, {
            ["aria-label"]: "close",
            onClick: (event: React.MouseEvent<any>) => {
              if (onClick) {
                onClick(event);
              }
              onClose();
            },
            ref,
            ...rest,
          })
        }
      </ModalContext.Consumer>
    );
  },
  { as: "button" },
);
