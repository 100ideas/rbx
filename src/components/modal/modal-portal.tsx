import classNames from "classNames";
import * as React from "react";

import { Generic } from "src/base";
import { initialValue, ModalContext, ModalContextValue } from "./modal-context";

export interface ModalPortalModifierProps {
  as?: React.ReactType; // tslint:disable-line:no-reserved-keywords
  className?: string;
  close: ModalContextValue["close"];
  closeOnBlur?: ModalContextValue["closeOnBlur"];
  closeOnEsc?: ModalContextValue["closeOnEsc"];
  innerRef?: React.Ref<HTMLElement | keyof JSX.IntrinsicElements>;
}

export type ModalPortalProps = ModalPortalModifierProps;

export class ModalPortal extends React.PureComponent<ModalPortalProps> {
  public static defaultProps = initialValue;

  public componentDidMount() {
    document.addEventListener("keydown", this.handleKeydown);
  }

  public componentWillUnmount() {
    document.removeEventListener("keydown", this.handleKeydown);
  }

  public render() {
    const {
      className,
      close,
      closeOnBlur,
      closeOnEsc,
      innerRef,
      ...rest
    } = this.props;

    return (
      <ModalContext.Provider
        value={{
          close,
          closeOnBlur: closeOnBlur === true,
          closeOnEsc: closeOnEsc === true,
        }}
      >
        <Generic
          className={classNames("modal", "is-active", className)}
          ref={innerRef}
          {...rest}
        />
      </ModalContext.Provider>
    );
  }

  private readonly handleKeydown = (event: KeyboardEvent) => {
    if (this.props.closeOnEsc === true && event.code === "Escape") {
      this.props.close();
    }
  }
}
