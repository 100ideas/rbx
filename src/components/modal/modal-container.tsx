import PropTypes from "prop-types";
import React from "react";
import ReactDOM from "react-dom";

import { refPropType } from "../../prop-types-extensions";
import { canUseDOM } from "../../utils";
import { ModalContextValue } from "./modal-context";
import { ModalPortal } from "./modal-portal";

export type ModalContainerProps = Partial<{
  active: boolean;
  as: React.ReactType<any>;
  children: React.ReactNode;
  closeOnBlur: ModalContextValue["closeOnBlur"];
  closeOnEsc: ModalContextValue["closeOnEsc"];
  containerClassName: string;
  innerRef: React.Ref<any>;
  onClose: () => void;
}>;

const propTypes = {
  active: PropTypes.bool,
  // as: PropType checked by Generic
  closeOnBlur: PropTypes.bool,
  closeOnEsc: PropTypes.bool,
  containerClassName: PropTypes.string,
  innerRef: refPropType,
  onClose: PropTypes.func,
};

export interface ModalContainerState {
  active: boolean;
}

export class ModalContainer extends React.PureComponent<
  ModalContainerProps,
  ModalContainerState
> {
  public static propTypes = propTypes;
  public readonly state: ModalContainerState = { active: false };
  private el: HTMLDivElement | undefined;

  constructor(props: ModalContainerProps) {
    super(props);
    this.state = { active: !!this.props.active };
    if (canUseDOM()) {
      this.el = document.createElement("div");
      if (props.containerClassName) {
        this.el.className = props.containerClassName;
      }
    }
  }

  public componentDidMount() {
    if (canUseDOM()) {
      document.body.appendChild(this.el!);
    }
  }

  public componentWillUnmount() {
    if (canUseDOM()) {
      document.body.removeChild(this.el!);
    }
  }

  public render() {
    const { active, containerClassName, ...rest } = this.props;
    return this.el && this.state.active
      ? ReactDOM.createPortal(
          <ModalPortal close={this.close} {...rest} />,
          this.el,
        )
      : null;
  }

  private close = () => {
    this.setState({ active: false });
    if (this.props.onClose) {
      this.props.onClose();
    }
  }
}
