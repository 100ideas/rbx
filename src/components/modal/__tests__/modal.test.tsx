import Enzyme from "enzyme";
import React from "react";

import { Modal } from "src/components/modal/modal";
import { ModalBackground } from "src/components/modal/modal-background";
import { ModalCard } from "src/components/modal/modal-card";
import { ModalClose } from "src/components/modal/modal-close";
import { ModalContainer } from "src/components/modal/modal-container";
import { ModalContent } from "src/components/modal/modal-content";
import { ModalContext } from "src/components/modal/modal-context";
import { ModalPortal } from "src/components/modal/modal-portal";

import {
  hasProperties,
  makeTestPropForwarding,
  validateBoolPropType,
  validatePropType,
  validateStringPropType,
} from "src/__tests__/testing";

const COMPONENT = Modal;
const DISPLAY_NAME = "Modal";
const DEFAULT_ELEMENT = "div";
// const BULMA_CLASS_NAME = "modal";

const testPropForwarding = makeTestPropForwarding(COMPONENT);

describe(`${DISPLAY_NAME} component`, () => {
  hasProperties(COMPONENT, {
    Background: ModalBackground,
    Card: ModalCard,
    Close: ModalClose,
    Container: ModalContainer,
    Content: ModalContent,
    Context: ModalContext,
    Portal: ModalPortal,
    defaultProps: { as: DEFAULT_ELEMENT },
  });

  test("it renders a ModalContainer", () => {
    const node = <Modal />;
    const wrapper = Enzyme.shallow(node);
    expect(wrapper.is(ModalContainer)).toBe(true);
  });

  describe("props", () => {
    const { propTypes } = COMPONENT;

    describe("active", () => {
      validateBoolPropType(propTypes, "active");

      [false, true].map(active => {
        testPropForwarding("active", active);
      });
    });

    describe("as", () => {
      testPropForwarding("as", "div");
    });

    describe("className", () => {
      testPropForwarding("className", "foo");
    });

    describe("closeOnBlur", () => {
      validateBoolPropType(propTypes, "closeOnBlur");

      [false, true].map(closeOnBlur => {
        testPropForwarding("closeOnBlur", closeOnBlur);
      });
    });

    describe("closeOnEsc", () => {
      validateBoolPropType(propTypes, "closeOnEsc");

      [false, true].map(closeOnEsc => {
        testPropForwarding("closeOnEsc", closeOnEsc);
      });
    });

    describe("containerClassName", () => {
      validateStringPropType(propTypes, "containerClassName");

      testPropForwarding("containerClassName", "foo");
    });

    describe("document", () => {
      validatePropType(propTypes, "document", [
        { value: document, valid: true, descriptor: "obj" },
        { value: "string", valid: false },
      ]);

      testPropForwarding("document", document);
    });

    describe("onClose", () => {
      validatePropType(propTypes, "onClose", [
        { value: () => undefined, valid: true, descriptor: "func" },
        { value: "string", valid: false },
      ]);

      testPropForwarding("onClose", () => undefined);
    });

    describe("ref", () => {
      testPropForwarding("ref", React.createRef(), "innerRef");
    });
  });
});
