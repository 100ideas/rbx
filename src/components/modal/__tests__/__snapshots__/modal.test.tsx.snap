// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should close the modal 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Modal Title
      </p>
      <button
        class="delete"
      />
    </header>
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should close the modal if clicked on background 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-content"
  >
    Content
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should close the modal on ESC key press 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Modal Title
      </p>
      <button
        class="delete"
      />
    </header>
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "Card": Object {
    "$$typeof": Symbol(react.forward_ref),
    "Body": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "section",
      },
      "render": [Function],
    },
    "Foot": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "footer",
      },
      "render": [Function],
    },
    "Head": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "header",
        "children": null,
        "showClose": true,
      },
      "render": [Function],
    },
    "Title": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "p",
      },
      "render": [Function],
    },
    "render": [Function],
  },
  "Content": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "div",
    },
    "render": [Function],
  },
  "render": [Function],
}
`;

exports[`Modal component should expose Content and Card's types 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "div",
  },
  "render": [Function],
}
`;

exports[`Modal component should expose Content and Card's types 2`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "Body": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "section",
    },
    "render": [Function],
  },
  "Foot": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "footer",
    },
    "render": [Function],
  },
  "Head": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "header",
      "children": null,
      "showClose": true,
    },
    "render": [Function],
  },
  "Title": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "p",
    },
    "render": [Function],
  },
  "render": [Function],
}
`;

exports[`Modal component should expose Content and Card's types 3`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "p",
  },
  "render": [Function],
}
`;

exports[`Modal component should expose Content and Card's types 4`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "header",
    "children": null,
    "showClose": true,
  },
  "render": [Function],
}
`;

exports[`Modal component should expose Content and Card's types 5`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "section",
  },
  "render": [Function],
}
`;

exports[`Modal component should expose Content and Card's types 6`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "footer",
  },
  "render": [Function],
}
`;

exports[`Modal component should not close the modal if clicked on background 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-content"
  >
    Content
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should not close the modal if clicked on background 2`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-content"
  >
    Content
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should not close the modal on ESC key press 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Modal Title
      </p>
      <button
        class="delete"
      />
    </header>
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should not close the modal on other than ESC key press 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Modal Title
      </p>
      <button
        class="delete"
      />
    </header>
    <section
      class="modal-card-body"
    >
      Body
    </section>
    <footer
      class="modal-card-foot"
    >
      Footer
    </footer>
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should not show close button 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-content"
  >
    Content
  </div>
</div>
`;

exports[`Modal component should not try to reopen if other prop change 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-content"
  >
    Content
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should open the modal 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Modal Title
      </p>
      <button
        class="delete"
      />
    </header>
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should render any child type 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
    role="presentation"
  />
  <div>
    CHILDREN
  </div>
  <div>
    CHILDREN
  </div>
  <button
    aria-label="close"
    class="modal-close is-large"
    type="button"
  />
</div>
`;

exports[`Modal component should render empty because no document on scope 1`] = `""`;
